File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/models/AnalyticsModels.kt
package ru.egetrainer.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.math.roundToInt

// MARK: student_analytics_screen_v1 (self / teacher scope)

@Serializable
data class AnalyticsScreen(
    val overall: OverallStats? = null,
    val student: AnalyticsStudent? = null,
    val sections: List<SectionStat>? = null,
    val topics: List<TopicStat>? = null,
    @SerialName("generated_at") val generatedAt: String? = null,
)

@Serializable
data class Counter(
    val total: Int,
    val correct: Int,
) {
    val pct: Int?
        get() = if (total > 0) (correct.toDouble() / total * 100).roundToInt() else null

    val ratioText: String get() = "$correct/$total"
}

@Serializable
data class OverallStats(
    val last3: Counter? = null,
    val last10: Counter? = null,
    val period: Counter? = null,
    @SerialName("all_time") val allTime: Counter? = null,
    @SerialName("last_seen_at") val lastSeenAt: String? = null,
)

@Serializable
data class AnalyticsStudent(
    val days: Int? = null,
    val grade: Int? = null,
    val source: String? = null,
    @SerialName("student_id") val studentId: String? = null,
    @SerialName("display_name") val displayName: String? = null,
    @SerialName("last_seen_at") val lastSeenAt: String? = null,
    @SerialName("viewer_scope") val viewerScope: String? = null,
)

@Serializable
data class SectionStat(
    @SerialName("section_id") val sectionId: String,
    val title: String? = null,
    val last10: Counter? = null,
    val period: Counter? = null,
    @SerialName("all_time") val allTime: Counter? = null,
    val coverage: Coverage? = null,
    @SerialName("last_seen_at") val lastSeenAt: String? = null,
) {
    val id: String get() = sectionId
}

@Serializable
data class Coverage(
    val pct: Int? = null,
    @SerialName("unics_total") val unicsTotal: Int? = null,
    @SerialName("unics_attempted") val unicsAttempted: Int? = null,
)

@Serializable
data class TopicStat(
    @SerialName("topic_id") val topicId: String,
    @SerialName("section_id") val sectionId: String? = null,
    @SerialName("subtopic_id") val subtopicId: String? = null,
    val title: String? = null,
    @SerialName("topic_order") val topicOrder: Int? = null,
    val last3: Counter? = null,
    val last10: Counter? = null,
    val period: Counter? = null,
    @SerialName("all_time") val allTime: Counter? = null,
    val coverage: Coverage? = null,
    val derived: DerivedStates? = null,
    @SerialName("last_seen_at") val lastSeenAt: String? = null,
    @SerialName("subtopic_last3_avg_pct") val subtopicLast3AvgPct: Double? = null,
) {
    val id: String get() = topicId
}

/** Производные состояния подтемы (бейджи «слабое», «давно не решал» и т.п.). */
@Serializable
data class DerivedStates(
    @SerialName("sample_state") val sampleState: String? = null,       // low | enough
    @SerialName("coverage_state") val coverageState: String? = null,   // covered | partial | none
    @SerialName("freshness_state") val freshnessState: String? = null, // fresh | stale
    @SerialName("performance_state") val performanceState: String? = null, // weak | mid | strong
)

File: ios/EGETrainerApp/EGETrainerApp/Models/AnalyticsModels.swift
import Foundation

// MARK: - student_analytics_screen_v1 (self / teacher scope)

struct AnalyticsScreen: Decodable {
    var overall: OverallStats?
    var student: AnalyticsStudent?
    var sections: [SectionStat]?
    var topics: [TopicStat]?
    var generatedAt: String?

    enum CodingKeys: String, CodingKey {
        case overall, student, sections, topics
        case generatedAt = "generated_at"
    }
}

struct Counter: Decodable {
    var total: Int
    var correct: Int

    var pct: Int? {
        guard total > 0 else { return nil }
        return Int((Double(correct) / Double(total) * 100).rounded())
    }

    var ratioText: String { "\(correct)/\(total)" }
}

struct OverallStats: Decodable {
    var last3: Counter?
    var last10: Counter?
    var period: Counter?
    var allTime: Counter?
    var lastSeenAt: String?

    enum CodingKeys: String, CodingKey {
        case last3, last10, period
        case allTime = "all_time"
        case lastSeenAt = "last_seen_at"
    }
}

struct AnalyticsStudent: Decodable {
    var days: Int?
    var grade: Int?
    var source: String?
    var studentId: String?
    var displayName: String?
    var lastSeenAt: String?
    var viewerScope: String?

    enum CodingKeys: String, CodingKey {
        case days, grade, source
        case studentId = "student_id"
        case displayName = "display_name"
        case lastSeenAt = "last_seen_at"
        case viewerScope = "viewer_scope"
    }
}

struct SectionStat: Decodable, Identifiable {
    var sectionId: String
    var title: String?
    var last10: Counter?
    var period: Counter?
    var allTime: Counter?
    var coverage: Coverage?
    var lastSeenAt: String?

    enum CodingKeys: String, CodingKey {
        case title, last10, period, coverage
        case sectionId = "section_id"
        case allTime = "all_time"
        case lastSeenAt = "last_seen_at"
    }

    var id: String { sectionId }
}

struct Coverage: Decodable {
    var pct: Int?
    var unicsTotal: Int?
    var unicsAttempted: Int?

    enum CodingKeys: String, CodingKey {
        case pct
        case unicsTotal = "unics_total"
        case unicsAttempted = "unics_attempted"
    }
}

struct TopicStat: Decodable, Identifiable {
    var topicId: String
    var sectionId: String?
    var subtopicId: String?
    var title: String?
    var topicOrder: Int?
    var last3: Counter?
    var last10: Counter?
    var period: Counter?
    var allTime: Counter?
    var coverage: Coverage?
    var derived: DerivedStates?
    var lastSeenAt: String?
    var subtopicLast3AvgPct: Double?

    enum CodingKeys: String, CodingKey {
        case title, last3, last10, period, coverage, derived
        case topicId = "topic_id"
        case sectionId = "section_id"
        case subtopicId = "subtopic_id"
        case topicOrder = "topic_order"
        case allTime = "all_time"
        case lastSeenAt = "last_seen_at"
        case subtopicLast3AvgPct = "subtopic_last3_avg_pct"
    }

    var id: String { topicId }
}

/// Производные состояния подтемы (бейджи «слабое», «давно не решал» и т.п.).
struct DerivedStates: Decodable {
    var sampleState: String?       // low | enough
    var coverageState: String?     // covered | partial | none
    var freshnessState: String?    // fresh | stale
    var performanceState: String?  // weak | mid | strong

    enum CodingKeys: String, CodingKey {
        case sampleState = "sample_state"
        case coverageState = "coverage_state"
        case freshnessState = "freshness_state"
        case performanceState = "performance_state"
    }
}
